core/Tracker/Model.php

```diff
@@ -35,8 +43,19 @@ public function createAction($visitAction)
         return $id;
     }
 
+    /**
+     * Write a goal conversion to the database
+     *
+     * @param array $conversion
+     *
+     * @return bool
+     * @throws Db\DbException
+     */
     public function createConversion($conversion)
     {
+        // Calculate and store the pages viewed before this conversion
+        $conversion['pageviews_before'] = $this->getPageCountViewedBeforeDate($conversion['idvisit'], $conversion['server_time']);
```

Comment:
Any reason for not handling this and also the database column stuff using a ConversionDimension class like https://github.com/matomo-org/matomo/blob/5.x-dev/plugins/ExampleTracker/Columns/ExampleConversionDimension.php#L27

Reply:
A couple of reasons:

- The `onGoalConversion` event doesn't provide access to the conversion server_time field needed for the query. Using the current time could theoretically cause inaccuracy on a slow server if the conversion server time was say 23:00:00.999, an additional (parallel) action was performed at 23:00:01.200 and the `onGoalConversion` runs at 23:00:01.400, then an additional pageview would be incorrectly counted. 

- ConversionDimension classes don't usually perform additional queries and the tracker db object is not exposed for the event. It would be more efficient to perform the query using the existing tracker db connection.

The performance and accuracy gains might well be negligible, and it would be good to be consistent. So I'm happy to rework this to use a dimension class if you think that's a better approach in this case?

Reply:
Great idea @sgiehl, #17466 seems like an excellent way to avoid the extra query and a useful addition on it's own :+1:

Reply:
Indeed we likely wouldn't want that query as the tracking part would cause more load then the query. Like some other fields can we maybe copy over the `pageview_position` from the log_link_visit_action? We may have this information already in some cases already?
Potentially another way might be `log_visit.visit_total_interactions - log_visit.visit_total_searches`?

Reply:
All other fields also contain events or other type of actions. But we actually only want to know the page views. The other PR would add a dimension for exactly that purpose, but would mean a longer running migration though.

Reply:
Just to confirm: `log_visit.visit_total_interactions - log_visit.visit_total_searches` should give us the amount of page views? Or are you saying it includes events? From what I see it should be the same as what is implemented in #17466 without needing a migration?

Reply:
https://github.com/matomo-org/matomo/blob/aea8f07de8ea3ce6ce39d3a217af5ec6a957db96/plugins/Actions/Columns/VisitTotalInteractions.php#L92

From what I can see the `VisitTotalInteractions` metric only counts actions if they have a URL or are `Action::TYPE_SITE_SEARCH` so `log_visit.visit_total_interactions - log_visit.visit_total_searches` looks like it will work. :slightly_smiling_face: 

Only possible issue is that that visitInfo values to be incremented get overwritten by other request processors so the `visitTotalInteractions` vlaue will be `'visit_total_interactions + 1'` instead of the original numeric value. I can add an option to the `VisitProperties` class to store a read only copy of the original visitInfo values to work around this.